astropy/coordinates/baseframe.py

```diff
@@ -639,19 +645,60 @@ def __init__(self, *args, **kwargs):
             raise TypeError(
                 '{0}.__init__ had {1} remaining unhandled arguments'.format(
                     self.__class__.__name__, len(args)))
+
+        self._data = representation_data
+
+        values = {}
+        for fnm, fdefault in self.get_frame_attr_names().items():
+            # Read-only frame attributes are defined as FrameAttribue
+            # descriptors which are not settable, so set 'real' attributes as
+            # the name prefaced with an underscore.
+
+            if fnm in kwargs:
+                value = kwargs.pop(fnm)
+                setattr(self, '_' + fnm, value)
+                # Validate attribute by getting it.  If the instance has data,
+                # this also checks its shape is OK.  If not, we do it below.
+                values[fnm] = getattr(self, fnm)
+            else:
+                setattr(self, '_' + fnm, fdefault)
+                self._attr_names_with_defaults.append(fnm)
+
         if kwargs:
             raise TypeError(
                 'Coordinate frame got unexpected keywords: {0}'.format(
                     list(kwargs)))
 
-        self._data = representation_data
-
-        # We do ``is not None`` because self._data might evaluate to false for
+        # We do ``is None`` because self._data might evaluate to false for
         # empty arrays or data == 0
-        if self._data is not None:
+        if self._data is None:
+            # No data: we still need to check that any non-scalar attributes
+            # have consistent shapes.
+            shapes = {fnm: value.shape for fnm, value in values.items()
+                      if getattr(value, 'size', 1) > 1}
+            if shapes:
+                if len(shapes) > 1:
+                    try:
+                        self._no_data_shape = check_broadcast(*shapes.values())
+                    except ValueError:
+                        raise ValueError(
+                            "non-scalar attributes with inconsistent "
+                            "shapes: {0}".format(shapes))
+
+                    for fnm in shapes:
```

Comment:
There are really two checks done: first, whether the shapes can be broadcast against each other; second, whether the attributes actually can be broadcast (this is true for all those that we define, but may not generally be the case). This is similar to the check done for the case that the frame does have data. (I'll again add a comment, since none of this is all that obvious...)